From a8cd89efe40e951937c24612f82e41744929163a Mon Sep 17 00:00:00 2001
From: Wen Gong <quic_wgong@quicinc.com>
Date: Fri, 23 Sep 2022 06:03:18 -0400
Subject: [PATCH 110/112] wifi: ath12k: send vdev down before vdev stop

Made changes to send vdev_down event before vdev_stop
in ath12k_bss_disassoc()

Signed-off-by: Wen Gong <quic_wgong@quicinc.com>
Signed-off-by: Amutha Ravikumar <quic_aravikum@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -3957,7 +3957,8 @@ static void ath12k_bss_assoc(struct ath1
 }
 
 static void ath12k_bss_disassoc(struct ath12k *ar,
-				struct ath12k_link_vif *arvif)
+				struct ath12k_link_vif *arvif,
+				bool do_vdev_down)
 {
 	struct ath12k_vif *tx_ahvif;
 	struct ath12k_link_vif *tx_arvif;
@@ -3966,6 +3967,12 @@ static void ath12k_bss_disassoc(struct a
 
 	lockdep_assert_held(&ar->conf_mutex);
 
+	if (do_vdev_down) {
+		ath12k_dbg(ar->ab, ATH12K_DBG_MAC, "mac vdev %i disassoc bssid %pM\n",
+			   arvif->vdev_id, arvif->bssid);
+		return;
+	}
+
 	ath12k_dbg(ar->ab, ATH12K_DBG_MAC, "mac vdev %i disassoc bssid %pM\n",
 		   arvif->vdev_id, arvif->bssid);
 
@@ -4518,7 +4525,7 @@ static void ath12k_mac_bss_info_changed(
 		if (vif->cfg.assoc)
 			ath12k_bss_assoc(ar, arvif, info);
 		else
-			ath12k_bss_disassoc(ar, arvif);
+			ath12k_bss_disassoc(ar, arvif, false);
 	}
 
 	if (changed & BSS_CHANGED_TXPOWER) {
@@ -4675,7 +4682,7 @@ static void ath12k_mac_op_vif_cfg_change
 		if (vif->cfg.assoc)
 			ath12k_bss_assoc(ar, arvif, &vif->bss_conf);
 		else
-			ath12k_bss_disassoc(ar, arvif);
+			ath12k_bss_disassoc(ar, arvif, false);
 	}
 }
 
@@ -12627,6 +12634,8 @@ ath12k_mac_op_unassign_vif_chanctx(struc
 	}
 
 stop_primary:
+	ath12k_bss_disassoc(ar, arvif, true);
+
 	ret = ath12k_mac_vdev_stop(arvif);
 	if (ret)
 		ath12k_warn(ab, "failed to stop vdev %i: %d\n",
